symsrv-fetch.py

```diff
@@ -24,345 +24,468 @@
-        return None
-    path = os.path.join(tmpdir, debug_file[:-1] + '_')
-    with open(path, 'wb') as f:
-        f.write(data)
-    try:
-        # Decompress it
-        subprocess.check_call(['cabextract', '-L', '-d', tmpdir, path],
-                              stdout=open(os.devnull, 'wb'))
-        os.unlink(path)
-        return os.path.join(tmpdir, debug_file.lower())
-    except subprocess.CalledProcessError:
-        return None
-
-
-def fetch_and_dump_symbols(tmpdir, debug_id, debug_file,
-                           code_id=None, code_file=None):
-    pdb_path = fetch_symbol_and_decompress(tmpdir, debug_id, debug_file)
-    if pdb_path is None:
-        return None
-    bin_path = None
-    while True:
+async def server_has_file(client, server, filename):
+    """
+    Send the symbol server a HEAD request to see if it has this symbol file.
+    """
+    url = urljoin(server, quote(filename))
+    for _ in range(10):
         try:
-            # Dump it
-            syms = subprocess.check_output(config.dump_syms_cmd + [pdb_path])
-            lines = syms.splitlines()
-            if not lines:
-                return None
-            os.unlink(pdb_path)
-            if bin_path:
-                os.unlink(bin_path)
-            return syms
-        except subprocess.CalledProcessError as e:
-            lines = e.output.splitlines()
-            bits = lines[0].split() if len(lines) > 0 else []
-            cpu = bits[2] if len(bits) > 2 else ''
-            if bin_path is None and cpu == 'x86_64':
-                # Can't dump useful symbols for Win64 PDBs without binaries.
-                if code_id and code_file:
-                    log.debug('Fetching binary %s, %s', code_id, code_file)
-                    bin_path = fetch_symbol_and_decompress(tmpdir, code_id, code_file)
-                    if bin_path:
-                        log.debug('Fetched binary %s', bin_path)
+            async with client.head(url, headers=HEADERS, allow_redirects=True) as resp:
+                if resp.status == 200 and (
+                    (
+                        "microsoft" in server
+                        and resp.headers["Content-Type"] == "application/octet-stream"
+                    )
+                    or "mozilla" in server
+                ):
+                    log.debug(f"File exists: {url}")
+                    return True
+                else:
+                    return False
+        except Exception as e:
+            log.debug(f"Error with {url}: retry")
+            log.exception(e)
+            await asyncio.sleep(0.5)
+
+    log.debug(f"Too much retries (HEAD) for {url}: give up.")
+    return False
+
+
+async def fetch_file(client, server, filename):
+    """
+    Fetch the file from the server
+    """
+    url = urljoin(server, quote(filename))
+    log.debug(f"Fetch url: {url}")
+    for _ in range(10):
+        try:
+            async with client.get(url, headers=HEADERS, allow_redirects=True) as resp:
```

Comment:
A possible improvement for the future: we could avoid downloading the whole file but just get the first few bytes and check the type and continue only when the type is OK. The task runtime is already very short though, so probably not worth it.

Reply:
I think it isn't worth doing it because most of the time the files are ok (pdb v2 are pretty rare nowadays and "unknown" is just here to handle an intermittent error I got locally)

Reply:
OK, makes sense.